2021/TurcoZiccolella/app.py

```diff
@@ -0,0 +1,491 @@
+                    talker.ResetTotalPeriodCounter()
+
+                first_cycle = False
+
+            cycle_in_sum = 0
+            cycle_out_sum = 0
+            cycle_both_sum = 0
+
+            # Reconstruct the file name
+            ts,f_name= pl.get_message() #Blocking 
+            timestring = datetime.datetime.fromtimestamp(ts).strftime(p)
+            print("nuovo file generato , aggiorno\n\n\n")
+
+            os.seteuid(0) #Enabling root -> we are reading
+
+            flow_streamer = NFStreamer(source=f_name, statistical_analysis=True, promiscuous_mode=True,
+                                       accounting_mode=0)
+
+            flows = list(flow_streamer) #temporary structure to avoid to hold root 
+            
+            os.remove(f_name)  # Remove temporary file 
+            os.seteuid(int(os.environ.get('SUDO_GID'))) #Disabling root
+
+
+            # Studying flows , if it's a new flow create the structure and rdd o/w update it
+            for flow in flows:
+
+                print(flow.application_name + " -> " + flow.src_ip + " -> " + flow.dst_ip + "[" + str(
+                    flow.src2dst_bytes) + "] " \
+                      + flow.dst_ip + " -> " + flow.src_ip + "[" + str(flow.dst2src_bytes) + "]")
+
+                if ipaddress.ip_address(flow.src_ip).is_private and ipaddress.ip_address(flow.dst_ip).is_private:
+                    continue
+
+                if filter_mode == "ip":
+
+                    if ipaddress.ip_address(flow.dst_ip).is_private:
+                        curr_talker = talkers.get(flow.src_ip) # try to get the talker associated with the key
+                        if curr_talker is None: #if it's None then we have to insert it
+
+                            if( ntalkers < max_talkers):
+                                # create talker structure
+                                curr_talker = Talker(flow.src_ip, flow.application_name,flow.src2dst_bytes,
+                                                    flow.dst2src_bytes, "ip"
+                                                    ,flow.application_category_name, flow.application_is_guessed,
+                                                    flow.requested_server_name, flow.client_fingerprint,
+                                                    flow.server_fingerprint,
+                                                    flow.user_agent, flow.content_type)
+
+                                # create his RRD only when talker is first seen
+                                curr_talker.RRDcreate(rrd_step, hb, min_rrd_value, max_rrd_value, ts, creationtimestring)
+                                # insert in talkers
+                                talkers[flow.src_ip] = curr_talker
+                                # new talker
+                                ntalkers += 1
+
+                        else:
+                            curr_talker.update(ts,flow.src2dst_bytes,flow.dst2src_bytes)
+
```

Comment:
Un nuovo talker viene aggiunto solo se è risulta vero l'if di riga 167, la creazione di un RRD avviene solo nel caso in cui viene individuato un nuovo talker, altrimenti si aggiornano soltanto i campi della struttura dati del talker

Reply:
Tutti i parametri vengono chiesti nella fase di setup() implementata in init_functions ed eseguita nel caso in cui il config fosse incompleto o l'utente abbia deciso di sovrascriverlo. Viene successivamente viene letto il config.ini , estratti i parametri e passati ai due thread specializzandone i metodi che eseguono.